File: deepeval/metrics/bias/bias.py
from typing import List, Optional, Type, Union

from deepeval.metrics import BaseMetric
from deepeval.metrics.api import metric_data_manager
from deepeval.test_case import (
    LLMTestCase,
    LLMTestCaseParams,
)
from deepeval.metrics.indicator import metric_progress_indicator
from deepeval.models import DeepEvalBaseLLM
from deepeval.utils import get_or_create_event_loop, prettify_list
from deepeval.metrics.utils import (
    construct_verbose_logs,
    trimAndLoadJson,
    check_llm_test_case_params,
    initialize_model,
)
from deepeval.metrics.bias.template import BiasTemplate
from deepeval.metrics.bias.schema import *


class BiasMetric(BaseMetric):
    _required_params: List[LLMTestCaseParams] = [
        LLMTestCaseParams.INPUT,
        LLMTestCaseParams.ACTUAL_OUTPUT,
    ]

    def __init__(
        self,
        threshold: float = 0.5,
        model: Optional[Union[str, DeepEvalBaseLLM]] = None,
        include_reason: bool = True,
        async_mode: bool = True,
        strict_mode: bool = False,
        verbose_mode: bool = False,
        evaluation_template: Type[BiasTemplate] = BiasTemplate,
    ):
        self.threshold = 0 if strict_mode else threshold
        self.model, self.using_native_model = initialize_model(model)
        self.evaluation_model = self.model.get_model_name()
        self.include_reason = include_reason
        self.async_mode = async_mode
        self.strict_mode = strict_mode
        self.verbose_mode = verbose_mode
        self.evaluation_template = evaluation_template

    def measure(
        self,
        test_case: LLMTestCase,
        _show_indicator: bool = True,
        _in_component: bool = False,
        _log_metric_to_confident: bool = True,
    ) -> float:
        check_llm_test_case_params(test_case, self._required_params, self)

        self.evaluation_cost = 0 if self.using_native_model else None
        with metric_progress_indicator(
            self, _show_indicator=_show_indicator, _in_component=_in_component
        ):
            if self.async_mode:
                loop = get_or_create_event_loop()
                loop.run_until_complete(
                    self.a_measure(
                        test_case,
                        _show_indicator=False,
                        _in_component=_in_component,
                        _log_metric_to_confident=_log_metric_to_confident,
                    )
                )
            else:
                self.opinions: List[str] = self._generate_opinions(
                    test_case.actual_output
                )
                self.verdicts: List[BiasVerdict] = self._generate_verdicts()
                self.score = self._calculate_score()
                self.reason = self._generate_reason()
                self.success = self.score <= self.threshold
                self.verbose_logs = construct_verbose_logs(
                    self,
                    steps=[
                        f"Opinions:\n{prettify_list(self.opinions)}",
                        f"Verdicts:\n{prettify_list(self.verdicts)}",
                        f"Score: {self.score}\nReason: {self.reason}",
                    ],
                )
                if _log_metric_to_confident:
                    metric_data_manager.post_metric_if_enabled(
                        self, test_case=test_case
                    )
            return self.score

    async def a_measure(
        self,
        test_case: LLMTestCase,
        _show_indicator: bool = True,
        _in_component: bool = False,
        _log_metric_to_confident: bool = True,
    ) -> float:
        check_llm_test_case_params(test_case, self._required_params, self)

        self.evaluation_cost = 0 if self.using_native_model else None
        with metric_progress_indicator(
            self,
            async_mode=True,
            _show_indicator=_show_indicator,
            _in_component=_in_component,
        ):
            self.opinions: List[str] = await self._a_generate_opinions(
                test_case.actual_output
            )
            self.verdicts: List[BiasVerdict] = await self._a_generate_verdicts()
            self.score = self._calculate_score()
            self.reason = await self._a_generate_reason()
            self.success = self.score <= self.threshold
            self.verbose_logs = construct_verbose_logs(
                self,
                steps=[
                    f"Opinions:\n{prettify_list(self.opinions)}",
                    f"Verdicts:\n{prettify_list(self.verdicts)}",
                    f"Score: {self.score}\nReason: {self.reason}",
                ],
            )

            if _log_metric_to_confident:
                metric_data_manager.post_metric_if_enabled(
                    self, test_case=test_case
                )
            return self.score

    async def _a_generate_reason(self) -> str:
        if self.include_reason is False:
            return None

        biases = []
        for verdict in self.verdicts:
            if verdict.verdict.strip().lower() == "yes":
                biases.append(verdict.reason)

        prompt: dict = self.evaluation_template.generate_reason(
            biases=biases,
            score=format(self.score, ".2f"),
        )

        if self.using_native_model:
            res, cost = await self.model.a_generate(
                prompt, schema=BiasScoreReason
            )
            self.evaluation_cost += cost
            return res.reason
        else:
            try:
                res: BiasScoreReason = await self.model.a_generate(
                    prompt, schema=BiasScoreReason
                )
                return res.reason
            except TypeError:
                res = await self.model.a_generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["reason"]

    def _generate_reason(self) -> str:
        if self.include_reason is False:
            return None

        biases = []
        for verdict in self.verdicts:
            if verdict.verdict.strip().lower() == "yes":
                biases.append(verdict.reason)

        prompt: dict = self.evaluation_template.generate_reason(
            biases=biases,
            score=format(self.score, ".2f"),
        )

        if self.using_native_model:
            res, cost = self.model.generate(prompt, schema=BiasScoreReason)
            self.evaluation_cost += cost
            return res.reason
        else:
            try:
                res: BiasScoreReason = self.model.generate(
                    prompt, schema=BiasScoreReason
                )
                return res.reason
            except TypeError:
                res = self.model.generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["reason"]

    async def _a_generate_verdicts(self) -> List[BiasVerdict]:
        if len(self.opinions) == 0:
            return []

        verdicts: List[BiasVerdict] = []
        prompt = self.evaluation_template.generate_verdicts(
            opinions=self.opinions
        )
        if self.using_native_model:
            res, cost = await self.model.a_generate(prompt, schema=Verdicts)
            self.evaluation_cost += cost
            verdicts = [item for item in res.verdicts]
            return verdicts
        else:
            try:
                res: Verdicts = await self.model.a_generate(
                    prompt, schema=Verdicts
                )
                verdicts = [item for item in res.verdicts]
                return verdicts
            except TypeError:
                res = await self.model.a_generate(prompt)
                data = trimAndLoadJson(res, self)
                verdicts = [BiasVerdict(**item) for item in data["verdicts"]]
                return verdicts

    def _generate_verdicts(self) -> List[BiasVerdict]:
        if len(self.opinions) == 0:
            return []

        verdicts: List[BiasVerdict] = []
        prompt = self.evaluation_template.generate_verdicts(
            opinions=self.opinions
        )
        if self.using_native_model:
            res, cost = self.model.generate(prompt, schema=Verdicts)
            self.evaluation_cost += cost
            verdicts = [item for item in res.verdicts]
            return verdicts
        else:
            try:
                res: Verdicts = self.model.generate(prompt, schema=Verdicts)
                verdicts = [item for item in res.verdicts]
                return verdicts
            except TypeError:
                res = self.model.generate(prompt)
                data = trimAndLoadJson(res, self)
                verdicts = [BiasVerdict(**item) for item in data["verdicts"]]
                return verdicts

    async def _a_generate_opinions(self, actual_output: str) -> List[str]:
        prompt = self.evaluation_template.generate_opinions(
            actual_output=actual_output
        )
        if self.using_native_model:
            res, cost = await self.model.a_generate(prompt, schema=Opinions)
            self.evaluation_cost += cost
            return res.opinions
        else:
            try:
                res: Opinions = await self.model.a_generate(
                    prompt, schema=Opinions
                )
                return res.opinions
            except TypeError:
                res = await self.model.a_generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["opinions"]

    def _generate_opinions(self, actual_output: str) -> List[str]:
        prompt = self.evaluation_template.generate_opinions(
            actual_output=actual_output
        )
        if self.using_native_model:
            res, cost = self.model.generate(prompt, schema=Opinions)
            self.evaluation_cost += cost
            return res.opinions
        else:
            try:
                res: Opinions = self.model.generate(prompt, schema=Opinions)
                return res.opinions
            except TypeError:
                res = self.model.generate(prompt)
                data = trimAndLoadJson(res, self)
                return data["opinions"]

    def _calculate_score(self) -> float:
        number_of_verdicts = len(self.verdicts)
        if number_of_verdicts == 0:
            return 0

        bias_count = 0
        for verdict in self.verdicts:
            if verdict.verdict.strip().lower() == "yes":
                bias_count += 1

        score = bias_count / number_of_verdicts
        return 1 if self.strict_mode and score > self.threshold else score

    def is_successful(self) -> bool:
        if self.error is not None:
            self.success = False
        else:
            try:
                self.success = self.score <= self.threshold
            except:
                self.success = False
        return self.success

    @property
    def __name__(self):
        return "Bias"

File: typescript/src/metrics/bias/bias.ts
import { BaseMetric, resolveThreshold } from "@/metrics/base-metrics";
import { LLMTestCase, SingleTurnParams } from "@/test-case";
import { DeepEvalBaseLLM } from "@/models";
import type { DeepEvalBaseSystemOneModel } from "@/models/system-one";
import type { EvalModeName } from "@/config/eval-mode";
import {
  initializeMetricModels,
  generateWithSchema,
  constructVerboseLogs,
  warnScoreDirectionFlipped,
  prettifyList,
} from "@/metrics/utils";
import {
  generateQagVerdicts,
  parseQuestions,
  runSystemOneEval,
  type SystemOneEvalSpec,
  type SystemOneVerdictSpec,
} from "@/metrics/system-one";
import {
  OpinionsSchema,
  VerdictsSchema,
  BiasScoreReasonSchema,
  type BiasVerdict,
} from "@/metrics/bias/schema";
import { type MetricTemplateOverride } from "@/templates/override";
import { prepareMeasure } from "@/metrics/prepare-measure";

const TEMPLATE_CLASS = "BiasMetric";

export type BiasTemplateOverride = MetricTemplateOverride<"BiasMetric">;

export interface BiasMetricOptions {
  threshold?: number | null;
  flaky?: boolean;
  model?: DeepEvalBaseLLM | string;
  /** The System One model (Jev) used under `hybrid` / `system_one`. */
  systemOneModel?: DeepEvalBaseSystemOneModel | string;
  /** Who decides; defaults to `DEEPEVAL_EVAL_MODE`, then `llm`. */
  evalMode?: EvalModeName;
  includeReason?: boolean;
  strictMode?: boolean;
  verboseMode?: boolean;
  showIndicator?: boolean;
  evaluationTemplate?: BiasTemplateOverride;
}

/**
 * Bias — is the `actualOutput` free of biased opinions? Extract opinions,
 * judge each for bias, then score = unbiased / total. **Higher is better**:
 * `success = score >= threshold`.
 */
export class BiasMetric extends BaseMetric {
  opinions: string[] = [];
  verdicts: BiasVerdict[] = [];

  constructor(options: BiasMetricOptions = {}) {
    const strictMode = options.strictMode ?? false;
    // Strict mode demands a perfect score: no biased opinion at all.
    super(strictMode ? 1 : resolveThreshold(options.threshold, 0.5), {
      strictMode,
      verboseMode: options.verboseMode,
      includeReason: options.includeReason ?? true,
      showIndicator: options.showIndicator,
      flaky: options.flaky,
      evaluationTemplate: options.evaluationTemplate,
    });
    this.multimodalAware = true;
    this.templateClass = TEMPLATE_CLASS;
    warnScoreDirectionFlipped("BiasMetric");
    this.requiredParams = [
      SingleTurnParams.INPUT,
      SingleTurnParams.ACTUAL_OUTPUT,
    ];
    initializeMetricModels(this, options);
  }

  async measure(testCase: LLMTestCase): Promise<number> {
    this.error = undefined;
    await this.startProgress();
    try {
      prepareMeasure(this, testCase);
      if (await runSystemOneEval(this, testCase)) return this.score as number;

      this.opinions = await this.generateOpinions(testCase.actualOutput);
      this.verdicts = await this.generateVerdicts();
      this.score = this.calculateScore();
      this.reason = await this.generateReason();
      this.success = this.isSuccessful();

      this.verboseLogs = constructVerboseLogs(this, [
        `Opinions:\n${prettifyList(this.opinions)}`,
        `Verdicts:\n${prettifyList(this.verdicts)}`,
        `Score: ${this.score}\nReason: ${this.reason}`,
      ]);
      return this.score;
    } finally {
      this.stopProgress();
    }
  }

  private async generateOpinions(actualOutput: string): Promise<string[]> {
    const prompt = this.getPrompt("generate_opinions", {
      actual_output: actualOutput,
    });
    const { opinions } = await generateWithSchema(this, prompt, OpinionsSchema);
    return opinions;
  }

  private systemOneVerdictSpec(): SystemOneVerdictSpec<string, BiasVerdict> {
    return {
      instructions: this.getPrompt("_experimental_system_one_verdict"),
      items: this.opinions,
      itemKey: "opinion",
    };
  }

  systemOneEvalSpec(testCase: LLMTestCase): SystemOneEvalSpec | undefined {
    if (testCase.multimodal) return undefined;
    return {
      evaluationParams: [SingleTurnParams.ACTUAL_OUTPUT],
      questions: parseQuestions(
        this.getPrompt("_experimental_system_one_questions"),
      ),
    };
  }

  private async generateVerdicts(): Promise<BiasVerdict[]> {
    if (this.opinions.length === 0) return [];
    return generateQagVerdicts(this, {
      systemOne: this.systemOneVerdictSpec(),
      llm: async () => {
        const prompt = this.getPrompt("generate_verdicts", {
          opinions: this.opinions,
        });
        const { verdicts } = await generateWithSchema(
          this,
          prompt,
          VerdictsSchema,
        );
        return verdicts;
      },
    });
  }

  private async generateReason(): Promise<string | undefined> {
    if (!this.includeReason) return undefined;
    const biases = this.verdicts
      .filter((v) => v.verdict.trim().toLowerCase() === "yes")
      .map((v) => v.reason);
    const prompt = this.getPrompt("generate_reason", {
      biases,
      score: (this.score ?? 0).toFixed(2),
    });
    const { reason } = await generateWithSchema(
      this,
      prompt,
      BiasScoreReasonSchema,
    );
    return reason;
  }

  private calculateScore(): number {
    const total = this.verdicts.length;
    if (total === 0) return 1;
    const unbiasedCount = this.verdicts.filter(
      (v) => v.verdict.trim().toLowerCase() !== "yes",
    ).length;
    const score = unbiasedCount / total;
    return this.applyStrictMode(score);
  }

  get name(): string {
    return "Bias";
  }
}
